corehq/apps/accounting/tests/test_invoice_pdf.py

```diff
@@ -0,0 +1,171 @@
+                invoice.canvas.setDateFormatter(empty)
+
+                # Prevent /ID metadata value
+                # HACK: Canvas doesn't expose a setter to customize this
+                invoice.canvas._doc.ID = empty
+
+                # render the PDF
+                invoice.get_pdf()
+            yield invoice_path, rendered
+
+    def iter_all_kwargs(self):
+        """Yields every combination of kwargs possible from
+        ``INVOICE_KWARG_CHOICES`` excluding those in ``INVALID_INVOICE_KWARGS``.
+        """
+        # NOTE: depends on the fact that python dicts are ordered
+        keys = list(self.INVOICE_KWARG_CHOICES)
+        for argv in product(*self.INVOICE_KWARG_CHOICES.values()):
+            kwargs = {keys[i]: v for i, v in enumerate(argv)}
+            if kwargs not in self.INVALID_INVOICE_KWARGS:
+                yield kwargs
+
+    def get_invoice(self, file_or_path, *, is_wire, is_customer, is_prepayment):
+        """Returns an invoice instantiated with all available args passed by
+        position (positionally?).
+
+        NOTE: the required kwargs of this function must match the keys of
+        ``INVOICE_KWARG_CHOICES``.
+        """
+        return InvoiceTemplate(
+            file_or_path,  # filename
+            self.TEST_LOGO_PATH,  # logo_filename
+            self.get_address("Tony"),  # from_address
+            self.get_address("Jono"),  # to_address
+            "group-therapy",  # project_name
+            date(2022, 5, 10),  # invoice_date
+            "42",  # invoice_number
+            "30",  # terms
+            date(2022, 6, 10),  # due_date
+            date(2022, 4, 1),  # date_start
+            date(2022, 4, 30),  # date_end
+            "$",  # bank_name
+            self.get_address("Paavo"),  # bank_address
+            "123",  # account_number
+            "456",  # routing_number_ach
+            "789",  # routing_number_wire
+            "---",  # swift_code
+            Decimal('-20.00'),  # applied_credit
+            Decimal('42.00'),  # subtotal
+            Decimal('6'),  # tax_rate
+            Decimal('2.52'),  # applied_tax
+            Decimal('24.52'),  # total
+            is_wire,  # is_wire
+            is_customer,  # is_customer
+            is_prepayment,  # is_prepayment
+            "Above & Beyond",  # account_name
+        )
+
+    @staticmethod
+    def get_address(name):
+        """Returns an address that renders well in any address part of an
+        Invoice PDF.
+        """
+        return Address(
+            name=name,
+            company_name="",
+            first_line=f"1 {name[::-1].title()} St",
+            second_line="",
+            city="Trance",
+            region="AB",
+            postal_code="300",
+            country="Around the World",
+            phone_number="",
+            email=f"{name.lower()}@example.com",
+            website="",  # populating this for a bank address messes up the footer
```

Comment:
Yeah, this method is used to create `Address` objects that the test uses for all of `From`, `To`, and `Bank` addresses. The report is built in such a way that if you add a website to a bank address, it will likely cause the bank footer to wrap, which garbles the doc. I'll improve this comment.